File: source/ts/src/astronomy/moonrise.ts
import { validateLocation } from '../utils/validation';
import { resolveEvent, type RiseSetKind } from './riseSetCache';
import type { GeoLocation } from '../types/location';

const LUNAR: RiseSetKind = { body: 'moon' };

/**
 * Next moonrise on or after the given UTC instant, or `null`, since some days have none. `limitDays` defaults to 2
 * (the Go port has no default). An Invalid Date, or a search that would reach 2^52 ms from 1970, throws `INVALID_DATE`.
 */
export function getMoonrise(
  searchFromUtc: Date,
  location: GeoLocation,
  limitDays: number = 2,
): Date | null {
  validateLocation(location);
  return resolveEvent(LUNAR, +1, searchFromUtc, location, limitDays);
}

/**
 * Next moonset on or after the given UTC instant, or `null` if none within `limitDays`, which defaults to 2 (the Go
 * port has no default). An Invalid Date, or a search that would reach 2^52 ms from 1970, throws `INVALID_DATE`.
 */
export function getMoonset(
  searchFromUtc: Date,
  location: GeoLocation,
  limitDays: number = 2,
): Date | null {
  validateLocation(location);
  return resolveEvent(LUNAR, -1, searchFromUtc, location, limitDays);
}

File: source/go/internal/astronomy/moonrise.go
package astronomy

import (
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var lunar = RiseSetKind{Body: RiseSetMoon}

func GetMoonrise(ctx *EphemerisCtx, searchFromMs int64, location types.GeoLocation, limitDays int) (int64, bool, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return 0, false, err
	}
	ms, ok := ResolveEvent(ctx, lunar, +1, searchFromMs, location, limitDays)
	return ms, ok, nil
}

func GetMoonset(ctx *EphemerisCtx, searchFromMs int64, location types.GeoLocation, limitDays int) (int64, bool, error) {
	if err := utils.ValidateLocation(location); err != nil {
		return 0, false, err
	}
	ms, ok := ResolveEvent(ctx, lunar, -1, searchFromMs, location, limitDays)
	return ms, ok, nil
}
